lib/logstash/inputs/file.rb

```diff
@@ -160,26 +164,71 @@ class LogStash::Inputs::File < LogStash::Inputs::Base
   # The default of 4095 is set in filewatch.
   config :max_open_files, :validate => :number
 
+  # What mode do you want the file input to operate in.
+  # Tail a few files or read many content-complete files
+  # The default is tail
+  # If "read" is specified then the following other settings are ignored
+  #   `start_position` (files are always read from the beginning)
+  #   `delimiter` (files are assumed to use \n or \r (or both) as line endings)
+  #   `close_older` (files are automatically 'closed' when EOF is reached)
+  # If "read" is specified then the following settings are heeded
+  #   `ignore_older` (older files are not processed)
+  # "read" mode now supports gzip file processing
+  config :mode, :validate => [ "tail", "read"], :default => "tail"
+
+  # When in 'read' mode, what action should be carried out when a file is done with.
+  # If 'delete' is specified then the file will be deleted.
+  # If 'log' is specified then the full path of the file is logged to the file specified
+  # in the `file_completed_log_path` setting.
+  config :file_completed_action, :validate => ["delete", "log"], :default => "delete"
```

Comment:
- could there be just a "do nothing" option? no delete, no log?
- also, are both options mutually exclusive?

Reply:
"do nothing" - I suppose so. This was a last minute change - people have asked for this since forever. I was going to generate a 'special' event tagged with EOF having the path and force people to handle it in their output section - IIRC Jordan suggested to not do the event thing and give people an action to take.

Doing nothing is problematic though, because if a user is reading 10000 files, say, the user does not know when the file is done with. The discover glob order is indeterminately OS specific. Do you think we should sort the watched file collection before the iteration? The order is then known and we can doc that. Then again with striped reading, files will be done with in an order that is determined by its size with the smaller files finishing before the larger files regardless of the glob order or the collection sorted order.


"do both" - yeah sure.

Reply:
It would be great to reorder files. 
It would be even better to have an option to tell how sort is done : by file name, by last modified file (from the oldest to the newest). 
I’d love the second option.

Reply:
@fbaligand 
Do you see yourself using striped reading?

Reply:
@guyboertje 
I just looked at "striped reading", if I understand well, it lets to read files in parallel, with a chunk read mechanism.
IMHO, it doesn't let to read files in a particular order. That feature is particularly useful when start_position=beginning, and at Logstash startup, there are already numerous files. It can be important that files are read using a specific order : from the oldest to the newest, or alphabetically.

Reply:
@fbaligand 
Sorry, my question was a general one, not related to discovered file collection order.

I am happy to add collection ordering, with a new settings `file_sort_by` -> "date"|"path" and `file_sort_direction` -> "asc"|"desc".
Q: do these ^ settings names and values make sense?

Reply:
That sounds great !
- Maybe replace "date" value by "last_modified" would be more meaningful ?
- Currently (so in tail mode), even if start_position=end, file input continues to discover new files every 15 seconds (by default) and read them from beginning. During these 15s, several files could arrive, that's why I find it relevant to apply order in tail mode.
- IMO, I prefer default as date + asc, as it means that files are processed as they arrived in time.

Reply:
@guyboertje 
I just looked at FileBeat options for that (https://www.elastic.co/guide/en/beats/filebeat/current/configuration-filebeat-options.html#scan-sort). 
They name it `scan.sort` and `scan.order`.
I like the prefix `scan_`, I find it meaningfull.